src/ucp/proto/proto_select.c

```diff
@@ -25,6 +28,221 @@ typedef struct {
+typedef struct {
+    size_t                         max_length;
+    ucp_proto_id_t                 proto_id;
+} ucp_proto_threshold_tmp_elem_t;
+
+
+UCS_ARRAY_DEFINE_INLINE(ucp_proto_thresh, unsigned,
+                        ucp_proto_threshold_tmp_elem_t);
+
+const ucp_proto_threshold_elem_t*
+ucp_proto_thresholds_search_slow(const ucp_proto_threshold_elem_t *thresholds,
+                                 size_t msg_length)
+{
+    unsigned idx;
+    for (idx = 0; msg_length > thresholds[idx].max_msg_length; ++idx);
+    return &thresholds[idx];
+}
+
+static ucs_status_t
+ucp_proto_thresholds_append(ucs_array_t(ucp_proto_thresh) *thresh_list,
+                            size_t max_length, ucp_proto_id_t proto_id)
+{
+    ucp_proto_threshold_tmp_elem_t *thresh_elem;
+    ucs_status_t status;
+
+    /* Consolidate with last protocol if possible */
+    if (!ucs_array_is_empty(thresh_list)) {
+        thresh_elem = ucs_array_last(thresh_list);
+        ucs_assertv(max_length > thresh_elem->max_length,
+                    "max_length=%zu last->max_length=%zu",
+                    max_length, thresh_elem->max_length);
+        if (thresh_elem->proto_id == proto_id) {
+            thresh_elem->max_length = max_length;
+            return UCS_OK;
+        }
+    }
+
+    status = ucs_array_append(ucp_proto_thresh, thresh_list);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    thresh_elem             = ucs_array_last(thresh_list);
+    thresh_elem->max_length = max_length;
+    thresh_elem->proto_id   = proto_id;
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_proto_thresholds_select_best(ucp_proto_id_mask_t proto_mask,
+                                 const ucs_linear_func_t *proto_perf,
+                                 ucs_array_t(ucp_proto_thresh) *thresh_list,
+                                 size_t start, size_t end)
+{
+    struct {
+        ucp_proto_id_t proto_id;
+        double         result;
+    } curr, best;
+    ucs_status_t status;
+    double x_intersect;
+    size_t midpoint;
+    char buf[64];
+
+    ucs_trace("candidate protocols for [%s]:",
+              ucs_memunits_range_str(start, end, buf, sizeof(buf)));
+    ucs_for_each_bit(curr.proto_id, proto_mask) {
+        ucs_trace("%24s %.0f+%.3f*X nsec",
+                  ucp_proto_id_field(curr.proto_id, name),
+                  proto_perf[curr.proto_id].c * UCS_NSEC_PER_SEC,
+                  proto_perf[curr.proto_id].m * UCS_NSEC_PER_SEC);
+    }
+
+    do {
+        ucs_assert(proto_mask != 0);
+
+        /* Find best protocol at the 'start' point */
+        best.result   = DBL_MAX;
+        best.proto_id = UCP_PROTO_ID_INVALID;
+        ucs_for_each_bit(curr.proto_id, proto_mask) {
+            curr.result = ucs_linear_func_apply(proto_perf[curr.proto_id],
+                                                start + 0.5);
+            if ((best.proto_id == UCP_PROTO_ID_INVALID) ||
+                (curr.result < best.result)) {
```

Comment:
if I'm not mistaken just checking `(curr.result < best.result)` should be enough, since `result < DBL_MAX`, isn't it?
also, we could add `ucs_assert(curr.result < DBL_MAX);`
wdyt?